alliance/src/ocp/doc/ocp.1
.\"   $Id: ocp.1,v 1.7 2006/03/30 19:07:45 xtof Exp $
.\" @(#)Labo.l 0.0 92/09/24 UPMC; Author: Christophe ALEXANDRE
.pl -.4
.TH OCP 1 "September 27, 2001" "ASIM/LIP6" "CAO\-VLSI Reference Manual"
.SH NAME
.TP
ocp
 - Standard Cell Placer
.so man1/alc_origin.1
.SH SYNOPSIS
.TP
\fBocp\fP 
[\fBoptions\fP] \fInetlist\fP \fIoutputname\fP
.br

.SH DESCRIPTION
\fBocp\fP is an automatic place tool for standard-cells.
.br

\fBinput net-list\fP
.br
The \fInetlist\fP file describes the input net-list.
.br
\fBocp\fP supports a hierarchical net-list. In this case the net-list is flattened by the placer according to the catalog file. The net-list format can be : structural VHDL, EDIF, or ALLIANCE internal format according to the environment variable \fBMBK_IN_LO\fP.
.br

\fBoutput layout\fP
.br
The file containing the placed block will have the name \fIoutputname\fP. This name is not optionnal and must always be present. The output format is defined by the environment variable \fBMBK_OUT_PH\fP.
.br

.SH OPTIONS
.br
\fBOptional cells placement file\fP
.br
\fB-partial <NAME>\fP
.br
A user defined pre-placement can be specified, thanks to a placement file.
.br
The optional placement file must be given by the user. It must have an extension that denotes the format defined by the environment variable \fBMBK_IN_PH\fP.
.br

\fBOptional connectors placement\fP
.br
The placement of connectors can be also specified.
.br
There are four ways to place connectors in a physical view :
.br
.br
\fB-partial <NAME>\fP pre-placed connectors defined by the user can be provided through the pre-placement file. In that case, automatic connector placement is impossible.
.br
\fB-c\fP option will automatically place the connectors randomly. The auto placement will set connectors on each side of the abutment box.
.br
\fB-ring\fP option will automatically place the connectors for the  ring pad placement tool. The placement is random but only on the north and south side unless the \fB-ioc <NAME>\fP option is given, in which case the constraints given in the <NAME>.ioc file will be respected, but with layers suitable for ring. 
.br
\fB-ioc <NAME>\fP option will place connectors as specified by the <NAME>.ioc given file.
.br

\fBMargin\fP
.br
It is possible to force the free area of the physical view.
.br
The \fB-margin <MARGIN>\fP option allow to set The amount of free area added in percentage of the cells area.
.br
The resulting area will be equal to CELL_AREA * (1 + <MARGIN>).
.br
By default, the margin value is 0.2 (20%)
.br
\fB-eqmargin\fP option will distribute equitably the margin between cells. By default, the biggest possible number of 2-pitch tie cells is inserted. This is done to have the best well and bulk polarity.
.br

\fBNumber of rows\fP
.br
\fB-rows <NR>\fP option forces the design to be placed in <NR> rows.
.br
The abutment box's width is automatically generated.
.br
This option won't be used if a defined placement file is given.
.br

\fBVerbose mode\fP
.br
\fB-v\fP set the verbose mode on
.br

\fBGnuplot\fP
.br
\fB-gnuplot\fP option allow to generate automatically gnuplot files, for editing statistics
.br
.SH IOC format
.PP
The IOC format is based on \fBCadence\fB input-output connectors placement description.
.br
The description is composed of 5 possibles sections: TOP(), BOTTOM(), LEFT(), RIGHT(), and IGNORE()
.br
In each section except IGNORE(), there are placed IOs.
.br
In the IGNORE() section, the IOs are ignored by \fBOCP\fB.
.br
.br
In every section, the IO syntax could be:
.br
for pin:       (IOPIN iopinName.0 );
.br
for space:     SPACE  value;                             
.br
The capital words are keywords. orientation is not required.
.br
The value is the space (number of pitches) between the IO above and the IO below it. 
.br
\fBExample\fB
.br
.nf
TOP ( # IOs are ordered from left to right
   (IOPIN b(3).0 );
   (IOPIN cin.0 );
   (IOPIN ck.0 );
   (IOPIN cout.0 );
)
BOTTOM ( # IOs are ordered from left to right
   (IOPIN i(3).0 );
   (IOPIN i(4).0 );
   (IOPIN i(5).0 );
   (IOPIN i(6).0 );
)

IGNORE ( # IOs are ignored(not placed) by IO Placer
)
.fi
.SH ENVIRONMENT VARIABLES
.PP
\fBocp\fP uses the environment variables MBK_VDD and MBK_VSS to know
the name of the power signals vdd and vss.
.br

.SH SEE ALSO
.PP
nero(1), MBK_IN_LO(1), MBK_IN_PH(1), MBK_OUT_PH(1), MBK_VDD(1), MBK_VSS(1)
.so man1/alc_bug_report.1


alliance/src/ocp/src/placer/Ocp.cpp

// This file is part of the Alliance Project.
// Copyright (C) Laboratoire LIP6 - Departement ASIM
// Universite Pierre et Marie Curie
//
// The Alliance Project  is free software;  you can  redistribute it and/or
// modify  it  under the  terms  of  the  GNU  General  Public License  as
// published by  the Free  Software Foundation; either  version 2  of  the
// License, or (at your option) any later version.
// 
// The Alliance Project  is distributed in the hope that it will be useful,
// but  WITHOUT  ANY  WARRANTY;  without  even  the  implied  warranty  of
// MERCHANTABILITY  or  FITNESS  FOR A  PARTICULAR PURPOSE.   See  the GNU
// General Public License for more details.
// 
// You should have received a copy  of  the  GNU  General  Public  License
// along with  the Alliance Project;  if  not,  write to the  Free Software
// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
//
//
// License-Tag
//
// Date   : 29/01/2004
// Author : Christophe Alexandre  <Christophe.Alexandre@lip6.fr>
//
// Authors-Tag 
#include <stdio.h>
#include <iostream>
#include <ctype.h>
using namespace std;

#include "mut.h"
#include "mph.h"
#include "mlo.h"

#include "PPlacement.h"

static void
Usage()
{
    cerr << "usage: ocp [options] <netlist> <filename>" << endl
         << "Options :"  << endl
         << "o -help : display man page" << endl
	 << "o -v : verbose mode" << endl
	 << "o -gnuplot : create statistics files to use with gnuplot" << endl
	 << "o -c : create connectors (placement randomly generated)" << endl
	 << "o -ring : create connectors suitable for ring pad placement tool" << endl
	 << "          placement randomly generated unless -ioc is given" << endl
	 << "o -ioc : create connectors and place it using .ioc file" << endl
         << "o -margin <MARGIN> : The amount of free area added " << endl
	 << "  in percentage of the cells area. " << endl
	 << "  The resulting area will be equal to " << endl
	 << "  CELL_AREA * (1 + <MARGIN>). Default value : 0.2" << endl
	 << "o -eqmargin : The margin is equitably distributed between all cells" << endl
	 << "  By default, we try to insert the maximum amount of 2-pitch tie cells" << endl
	 << "o -partial <PARTIAL> : " << endl
	 << "  The design is already partially placed (there must be a" << endl
	 << "  physical view describing this partial placement)." << endl
	 << "o -rows <NR> : the placement will have <NR> rows"<< endl
	 << "o -mdl <NL> : maximum loop of final optimisation "
	 << "default : <NL> = 2" << endl;
}

static void
PrintEnv()
{
    char* allianceTop = mbkgetenv("ALLIANCE_TOP");
    char* mbkInLo = mbkgetenv("MBK_IN_LO");
    char* mbkOutLo = mbkgetenv("MBK_OUT_LO");
    char* mbkInPh = mbkgetenv("MBK_IN_PH");
    char* mbkOutPh = mbkgetenv("MBK_OUT_PH");
    char* mbkVss = mbkgetenv("MBK_VSS");
    char* mbkVdd = mbkgetenv("MBK_VDD");
    char* mbkCatalName = mbkgetenv("MBK_CATAL_NAME");
    char* mbkCataLib = mbkgetenv("MBK_CATA_LIB");
    cout << "o ALLIANCE environment:" << endl;
    if (allianceTop)
	cout << "  o ALLIANCE_TOP   : " << allianceTop << endl;
    else
	cout << "  o Warning: no ALLIANCE_TOP variable set ...." << endl;
    cout << "o MBK environment:" << endl;
    if (mbkInLo)
	cout << "  o MBK_IN_LO      : " << mbkInLo << endl;
    else
	cout << "  o Warning: no MBK_IN_LO variable set ...." << endl;
    if (mbkOutLo)
	cout << "  o MBK_OUT_LO     : " << mbkOutLo << endl;
    else
	cout << "  o Warning: no MBK_OUT_LO variable set ...." << endl;
    if (mbkInPh)
	cout << "  o MBK_IN_PH      : " << mbkInPh << endl;
    else
	cout << "  o Warning: no MBK_IN_PH variable set ...." << endl;
    if (mbkOutPh)
	cout << "  o MBK_OUT_PH     : " << mbkOutPh << endl;
    else
	cout << "  o Warning: no MBK_OUT_PH variable set ...." << endl;
    if (mbkVss)
	cout << "  o MBK_VSS        : " << mbkVss << endl;
    else
	cout << "  o Warning: no MBK_VSS variable set ...." << endl;
    if (mbkVdd)
	cout << "  o MBK_VDD        : " << mbkVdd << endl;
    else
	cout << "  o Warning: no MBK_VDD variable set ...." << endl;
    if (mbkCatalName)
	cout << "  o MBK_CATAL_NAME : " << mbkCatalName << endl;
    else
	cout << "  o Warning: no MBK_CATAL_NAME variable set ...." << endl;
    if (mbkCataLib)
    {
	cout << "  o MBK_CATA_LIB   : ";
	while (*mbkCataLib !=  '\0')
	{
	    while ((*mbkCataLib != '\0') && (*mbkCataLib !=  ':'))
	    {
		cout << *mbkCataLib++;
	    }
	    if (*mbkCataLib == ':')
	    {
		cout << endl << "                     ";
		++mbkCataLib;
	    }
	    else
		cout << endl;
	}
    }
    else
	cout << "  o Warning: no MBK_CATA_LIB variable set ...." << endl;
    cout << endl;
}

int
main(int argc, char **argv)
{
    double		BBoxOccCostRatio	= 0.5;
    double		Margin			= 0.2;
    bool		marginflag		= false;
    int			MaxDetLoop		= 2;
    int			NbRows			= 0;
    char		PartialFile[256]	;	 //.ap file
    char		IocFile[256]		;	 //.ioc file
    bool		RowFlg			= false;
    bool		HelpFlg			= false; // display help
    bool		UsageFlg		= false; // print usage
    bool		PartialFlg		= false; // .ap file loaded
    bool		PlotFlg			= false; // GnuPlot graphics
    bool		ConFlg			= false; // Connectors
    bool		RingFlg			= false; // Connectors
    bool		IocFlg			= false; // file for Connectors
    bool		VerboseFlg		= false; // verbose mode
    bool		EqMarginFlg		= false; // don't try to maximise
    							 // 2-pitch cells
    double		NetMult			= 1.0 ;
    double		BinMult			= 0 ;
    double		RowMult			= 0 ;

    /* ###------------------------------------------------------### */
    /* analyse the command line, set option flags and find the	    */
    /* first file in the argument list				    */
    /* ###------------------------------------------------------### */

    mbkenv();
    
    alliancebanner ("OCP", VERSION, "Placer for Standards Cells", "2001", ALLIANCE_VERSION);

    if (argc < 2)
    {
	Usage();
	exit(0);
    }

    for (int i=1; i<argc; i++)
    {
	if (!strcmp (argv[i], "-help"))
	{
	    HelpFlg = true;
	}
	else
	    if (!strcmp (argv[i], "-v"))
	    {
		if (i+2 < argc) VerboseFlg = true;
		else	      UsageFlg = true;
	    }
	    else
		if (!strcmp (argv[i], "-eqmargin"))
		{
		    if (i+2 < argc) EqMarginFlg = true;
		    else	      UsageFlg = true;
		}
		else
		    if (!strcmp (argv[i], "-gnuplot"))
		    {
			if (i+2 < argc) PlotFlg = true;
			else	      UsageFlg = true;
		    }
		    else
			if (!strcmp (argv[i], "-c"))
			{
			    if (i+2 < argc) ConFlg = true;
			    else	      UsageFlg = true;
			}
			else
			    if (!strcmp (argv[i], "-ring"))
			    {
				if (i+2 < argc) RingFlg = true;
				else UsageFlg = true;
			    }
			    else
				if (!strcmp (argv[i], "-r"))
				{
				    if (i+3 < argc)
				    {
					if (!sscanf(argv[i+1], "%lg", &BBoxOccCostRatio))
					{
					    cout << "WARNING : invalid argument for -r" << endl;
					    UsageFlg = true;
					}
	      } else {
	    	cout << "WARNING : no file or argument given" << endl;
		UsageFlg = true;
	      }
            }

	    else
	   
            if (!strcmp (argv[i], "-nbrmult")) {
              if (i+5 < argc)
              {
        	if (!sscanf(argv[i+1], "%lg", &NetMult))
        	{
            	  cout << "WARNING : invalid argument for -nbrmult" << endl;
        	  UsageFlg = true;
        	}
        	if (!sscanf(argv[i+2], "%lg", &RowMult))
        	{
            	  cout << "WARNING : invalid argument for -nbrmult" << endl;
        	  UsageFlg = true;
        	}
        	if (!sscanf(argv[i+3], "%lg", &BinMult))
        	{
            	  cout << "WARNING : invalid argument for -nbrmult" << endl;
        	  UsageFlg = true;
        	}
              } else {
            	cout << "WARNING : no file or argument given" << endl;
        	UsageFlg = true;
        
              }
            }
	    
	    else
            
	    if (!strcmp (argv[i], "-margin")) {
	      if (i+3 < argc)
	      {
                if (!sscanf(argv[i+1], "%lg", &Margin))
		{
	    	  cout << "WARNING : invalid argument for -margin" << endl;
		  UsageFlg = true;
		}
	      } else {
	    	cout << "WARNING : no file or argument given" << endl;
		UsageFlg = true;
	      }
	      marginflag = true;
            }
	    
	    else
	    
	    if (!strcmp (argv[i], "-partial")) {
	      if (i+3 < argc)
	      {
                if (!sscanf(argv[i+1], "%s", PartialFile))
		{
	    	  cout << "WARNING : invalid argument for -partial" << endl;
		  UsageFlg = true;
		}
		PartialFlg = true;
	      } else {
	        cout << "WARNING : no file or argument given" << endl;
		UsageFlg = true;
	      }
            }
	    
	    else
	      
	    if (!strcmp (argv[i], "-ioc")) {
	      if (i+3 < argc)
	      {
                if (!sscanf(argv[i+1], "%s", IocFile))
		{
	    	  cout << "WARNING : invalid argument for -partial" << endl;
		  UsageFlg = true;
		}
		IocFlg = true;
		strcat(IocFile,".ioc");
	      } else {
	        cout << "WARNING : no file or argument given" << endl;
		UsageFlg = true;
	      }
            }
	    
	    else
	    
	    if (!strcmp (argv[i], "-mdl")) {
	      if (i+3 < argc)
	      {
                if (!sscanf(argv[i+1], "%d", &MaxDetLoop))
		{
		  cout << "WARNING : invalid argument for -mdl" << endl;
		  UsageFlg = true;
		}

	      } else {
		cout << "WARNING : no file or argument given" << endl;
		UsageFlg = true;
	      }
            }
	    
	    else
	    
	    if (!strcmp (argv[i], "-rows")) {
	      if (i+3 < argc)
	      {
	        if (!sscanf(argv[i+1], "%d", &NbRows))
		{
		  cout << "WARNING : invalid argument for -rows" << endl;
		  UsageFlg = true;
		}
		RowFlg = true;
	      } else {
		cout << "WARNING : no file or argument given" << endl;
		UsageFlg = true;
	      }
            }

	    else

	    if (!strncmp (argv[i], "-", 1))
	    {
	      cout << "WARNING : unavailable option : " << argv[i] << endl;
	      UsageFlg = true;
	    }
        }
  
    /* ###------------------------------------------------------### */
    /*      cases of error in the command line:			    */
    /*      - no argument					    */
    /*      - argument unrecognized as an option		    */
    /*      - number of files do not match:			    */
    /* ###------------------------------------------------------### */

    if (HelpFlg == true) {
      system("man ocp");
      exit (0);
    }

    if (argc < 3) {
      Usage();
      exit(1);
    }
    
    if ( BBoxOccCostRatio < 0.0 || BBoxOccCostRatio > 1.0)
        UsageFlg = true;

    if ( Margin < 0 )
        UsageFlg = true;
    
    if ( Margin < 0.15 )
    {
      cerr << " o Ocp Warning : the value of margin: "
	   << Margin << " is very low... Try a value around 0.15 ..." << endl;
    }

    if ( Margin > 3 )
    {
      cerr << " o Ocp Warning : the value of margin: "
	   << Margin << " cannot be respected..." << endl
	   << " o value forced to 0.4 !!" << endl; 
      Margin = 0.4;
    }

    if ( marginflag && PartialFlg )
    {
	cerr << " o Ocp Warning : the switch margin and partial are not coherent "
	     << "and should not be use together...." << endl
	     << " o In case of a preplacement the margin is determined"
	     << " by the value of the Abutment Box and the net netlist you give me." << endl
	     << " o In consequence the switch margin has no sense and is bypassed ...."
	     << endl;
    }

    if (IocFlg && ConFlg)
    {
	cerr << " o You cannot have a connectors placement file and ask "
	    << "for a automatic connectors placement at the same time ...." << endl
	    << " o Use either -ioc or -c" << endl;
	exit(1);
    }
    
    if (UsageFlg) {
        Usage();
        exit (1);
    }

    if (VerboseFlg)
	PrintEnv();

    // Initializations
    // ***************

    struct lofig* fig = NULL;
    fig = getlofig(argv[argc-2],'A');
    
    char *namefile = argv[argc-1];

    struct phfig* physicalfig = NULL;
    if (PartialFlg) physicalfig = getphfig(PartialFile,'A');

    // On fait le placement
    // ********************
    PPlacement Placement(ConFlg, RingFlg, RowMult, BinMult, NetMult,
	IocFlg, IocFile, PlotFlg, VerboseFlg, PartialFlg, EqMarginFlg,
	physicalfig, namefile);
    Placement.SetMargin(Margin);
    Placement.SetMaxDetLoop(MaxDetLoop);
    if (!RowFlg)
	NbRows = 0;
    Placement.Place(fig, NbRows);

    // On sauve le resultat
    // ********************
    cout  << endl << "Ocp : placement finished" << endl;
    if (Placement.GetBoolPlot())
    {
      Placement.PlotFinal("view");
      Placement.PlotStat();
      cout << "gnuplot files created - in order to use them, please type :"
	<< endl << "gnuplot afterglobal.gpl"
	<< endl << "gnuplot binsafterglobal.gpl"
	<< endl << "gnuplot init.gpl"
	<< endl << "gnuplot instances-init.gpl"
	<< endl << "gnuplot stat.gpl"
	<< endl << "gnuplot view.gpl" << endl;
    }
    int superr = Placement.Save();
    dellofig(fig->NAME);
    
    return (superr);
}
